lattices/src/semiring_application.rs

```diff
@@ -0,0 +1,226 @@
+    /// Create a new instance of Cost with the given value.
+    pub fn new(value: u32) -> Self {
+        Cost(value)
+    }
+}
+
+/// Implementation of the addition trait for Cost semiring.
+impl Addition<u32> for Cost {
+    /// Adds two elements of the semiring.
+    fn add(&self, input_1: &u32, input_2: &u32) -> u32 {
+        std::cmp::min(*input_1, *input_2)
+    }
+}
+
+/// Implementation of the multiplication trait for Cost semiring.
+impl Multiplication<u32> for Cost {
+    /// Multiplies two elements of the semiring.
+    fn mul(&self, input_1: &u32, input_2: &u32) -> u32 {
+        input_1.wrapping_add(*input_2)
+    }
+}
+
+/// Implementation of the Zero trait for Cost semiring.
+impl Zero<u32> for Cost {
+    /// Returns the zero element of the semiring.
+    fn zero(&self) -> u32 {
+        u32::MAX
+    }
+}
+
+/// Implementation of the One trait for Cost semiring.
+impl One<u32> for Cost {
+    /// Returns the one element of the semiring.
+    fn one(&self) -> u32 {
+        0
+    }
+}
+
+/// Implementation of the confidence Score semiring ([0, 1], max, *, 0, 1)
+pub struct ConfidenceScore(f64);
+
+impl ConfidenceScore {
+    /// Create a new instance of ConfidenceScore with the given value.
+    pub fn new(value: f64) -> Self {
+        // Ensure the value is within the range [0, 1]
+        let clamped_value = value.clamp(0.0, 1.0);
+        ConfidenceScore(clamped_value)
+    }
+}
+
+/// Implementation of the addition trait for ConfidenceScore semiring.
+impl Addition<f64> for ConfidenceScore {
+    /// Adds two elements of the semiring.
+    fn add(&self, input_1: &f64, input_2: &f64) -> f64 {
+        input_1.max(*input_2)
+    }
+}
+
+/// Implementation of the multiplication trait for ConfidenceScore semiring.
+impl Multiplication<f64> for ConfidenceScore {
+    /// Multiplies two elements of the semiring.
+    fn mul(&self, input_1: &f64, input_2: &f64) -> f64 {
+        *input_1 * *input_2
+    }
+}
+
+/// Implementation of the Zero trait for ConfidenceScore semiring.
+impl Zero<f64> for ConfidenceScore {
+    /// Returns the zero element of the semiring.
+    fn zero(&self) -> f64 {
+        0.0
+    }
+}
+
+/// Implementation of the One trait for ConfidenceScore semiring.
+impl One<f64> for ConfidenceScore {
+    /// Returns the one element of the semiring.
+    fn one(&self) -> f64 {
+        1.0
+    }
+}
+
+/// Implementation of Fuzzy Logic semiring ([0, 1], max, min, 0, 1).
+pub struct FuzzyLogic(f64);
+
+impl FuzzyLogic {
+    /// Create a new instance of FuzzyLogic with the given value.
+    pub fn new(value: f64) -> Self {
+        // Ensure the value is within the range [0, 1]
+        let clamped_value = value.clamp(0.0, 1.0);
+        FuzzyLogic(clamped_value)
+    }
+}
+
+/// Implementation of the addition trait for FuzzyLogic semiring.
+impl Addition<f64> for FuzzyLogic {
+    /// Adds two elements of the semiring.
+    fn add(&self, input_1: &f64, input_2: &f64) -> f64 {
+        input_1.max(*input_2)
```

Comment:
switch to cmp

Reply:
says that the trait `Ord` is not implemented for `f64`

https://stackoverflow.com/questions/69665188/min-max-of-vecf64-trait-ord-is-not-implemented-for-xy